src/main/java/net/spy/memcached/tls/TLSConnectionManager.java

```diff
@@ -0,0 +1,211 @@
+package net.spy.memcached.tls;
+
+import net.spy.memcached.compat.log.Logger;
+import net.spy.memcached.compat.log.LoggerFactory;
+
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLEngineResult.HandshakeStatus;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLSession;
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.SocketChannel;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+public class TLSConnectionManager implements Closeable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TLSConnectionManager.class);
+
+    private final SSLEngine sslEngine;
+
+    private SSLSession currentSession;
+
+    private ByteBuffer appOutBuffer; // Holds the data we are preparing to send out
+    private ByteBuffer appInBuffer; // Holds the data we have received and unwrapped
+    private ByteBuffer networkOutBuffer; // Holds the data we are sending across the wire
+    private ByteBuffer networkInBuffer; // Holds the data we have received from the wire
+
+    ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    public TLSConnectionManager(SSLContext sslContext) {
+        this.sslEngine = sslContext.createSSLEngine();
+
+        configureSslEngine();
+    }
+
+    private void configureSslEngine() {
+        // We are the client, not the server
+        sslEngine.setUseClientMode(true);
+    }
```

Comment:
Do we need 
```
engine.setEnabledProtocols(<protocol like  "TLSv1.2">);
engine.setEnabledCipherSuites(engine.getSupportedCipherSuites());
```
here too? not sure if I'm just getting ahead of things since the handshake already works, but I know those come into play in the handshake so wanted to check. Asking because they're used in the ES auth plugin, which I'd link normally but not sure I can here